### ISSUE TITLE
Refactor to add a utility to detect non-selector child nodes

### ISSUE THREAD

**Issue by romainmenke** [stylelint/stylelint#7915]
I suggest we add a generalized helper to detect when the contents of functional pseudo selectors (both classes and elements) aren't selectors.

Example:
- `selector-type-no-unknown` (`::highlight(bar)`)
- `selector-max-type` (`:active-view-transition(foo)`)
- `selector-max-class` (`::view-transition-group(*.foo)`)
- ...

Having a shared utility would make it easier to have consistent support and it would reduce the maintenance burden of updating these exceptions.

-----

I am not sure what a really good name would be for such a utility.

I considered `ignoredSelectorChildren` but I am not sure if we always want to `ignore`.
This name also doesn't explain why they are ignored.

Maybe
- `selectorArgumentsAreNotSelectors`
- `nonRecursiveFunctionalSelector`
- ...

naming things... :)

**Comment by Mouvedia** [comment]
> This function would take a `postcss-selector-parser` node and return `true` when it is one of the functional pseudo's that doesn't take a selector as its argument.

Wouldn't an util that check the argument itself be easier to maintain?
i.e. it will essentially be a black list / white list (whichever is smaller)

Let's take `:-moz-element()` as an example.
I wouldn't be surprised if someone forgot to add it to the white list.
`:-moz-element()` takes ids as arguments.
On the other hand, with a hypothetical `isSelector` utility it would be unthinkable to forget to check ids.
tl;dr: checking the argument would be more future proof IMHO but maybe I am missing the point

> I am not sure what a really good name would be for such a utility.

Don't bother.
Names depends on the usage.
i.e. should be done during the PR review

**Comment by Mouvedia** [comment]
> This function would take a `postcss-selector-parser` node and return `true` when it is one of the functional pseudo's that doesn't take a selector as its argument.

Wouldn't an util that check the argument itself be easier to maintain?
i.e. it will essentially be a black list or a white list (whichever is smaller)


I wouldn't be surprised if we forgot to add functions to the white/black list.
On the other hand, with a hypothetical `isSelector` utility it would be unthinkable for it to be buggy.

> I am not sure what a really good name would be for such a utility.

Don't bother.
Names depends on the usage.
i.e. should be done during the PR review

**Comment by Mouvedia** [comment]
Here's a non-exhaustive list of the ones not covered by the set added in #7913:

- `::part()`
- `:active-view-transition-type()`
- `:dir()`
-  `:-moz-locale-dir()`
- `:state()`
- `:lang()`
- `nth-*()`

**Comment by romainmenke** [comment]
We should also bikeshed this list: 

https://github.com/stylelint/stylelint/blob/b21389101495fcc9d184277862d06403f2503346/lib/reference/selectors.cjs#L185-L191

These are not pseudo elements that accept a custom idents, they are pseudo elements that accept arguments that aren't other selectors.

`.foo`, `*` are also valid arguments for the view transition related pseudo's and those aren't custom idents.

**Comment by Mouvedia** [comment]
This is off-topic in this issue. Can you please bikeshed the name in the PR?

**Comment by romainmenke** [comment]
We can bikeshed in the PR, but I currently have no idea what a good name would be.
The PR would be blocked until someone does come up with a good name.
I don't mind letting that PR remain blocked until a good name has been decided upon.

Naming these things needs to be done for the utility anyway and these names will be closely linked.

**Comment by romainmenke** [comment]
`:nth-*()` should not be part of this utility.

These do take selectors as their arguments in some cases.
The proposed utility should only cover those selectors that never take selectors as arguments.

I think `:nth-*()` is better handled by dedicated utilities.

**Comment by romainmenke** [comment]
🤔 or we flip it around and create a utility to walk a `postcss-selector-parser` AST and only visit real selectors?

- `li:nth-child(-n + 3 of .foo)` skips `-n + 3 of` while still visiting `.foo`
- `::view-transition-group(*.foo)` skips `*.foo`

I think that that is maybe a better boundary?

**Comment by Mouvedia** [comment]
off-topic

> We can bikeshed in the PR, but I currently have no idea what a good name would be.

The name is gone.

**Comment by ybiquitous** [comment]
Adding a utility sounds good to me. 👍🏼 

If you have no big concerns, please open a pull request. If we don't like the function name so much, we can rename it after merging the PR.

(Basically, I welcome pull requests to help us understand/read the code base more easily. Please don't hesitate to open them. 😄 )

**Comment by jeddy3** [comment]
> we flip it around and create a utility to walk a postcss-selector-parser AST and only visit real selectors?
> I think that that is maybe a better boundary?

SGTM, definitely worth exploring.
